subprojects/core/src/main/java/org/gradle/internal/service/scopes/ExecutionGlobalServices.java

```diff
@@ -188,17 +197,18 @@ TypeAnnotationMetadataStore createAnnotationMetadataStore(CrossBuildInMemoryCach
     InspectionSchemeFactory createInspectionSchemeFactory(
         List<TypeAnnotationHandler> typeHandlers,
         List<PropertyAnnotationHandler> propertyHandlers,
+        List<MethodAnnotationHandler> methodHandlers,
```

Comment:
🤔 I was not aware you could inject a list of multiple implementations of an interface like this.

But despite it being possible, is there a good reason for doing it (I understand this is copying the existing pattern for `PropertyAnnotationHandler`)?  Why not create these inside of `createInspectionSchemeFactory` instead of registering additional services which won't/can't be referenced individually outside of creating the factory here?  This existing code seems like it's needlessly throwing a bunch of stuff into the global service space.

Reply:
I think the best reason is that we can construct the inspection scheme factory without creating a direct dependency on the project providing the annotation handler.  In other words, we reduce the coupling between `core` (or wherever this service lives in the future) and the projects where all of the annotation handlers are.  I imagine this also avoids some weird cycles, too.

Rather than pull all of this stuff into one place, I'd rather think about how we can leave it uncoupled but make it easier to understand what annotations/handlers are used in what schemes and reduce the boilerplate necessary to register a new annotation.  For instance, maybe we could add an annotation to the handler objects that says it is a "task scheme" annotation.  Then, when we are instantiating the task scheme, we get all of the annotation handlers with the task scheme and have them tell us the types they handle.  Then, creating a new annotation only involves adding a handler with the right scheme annotation, and if you wanted to know all of the types involved in the task scheme, you could simply search for references to the scheme.  Just one idea, I'm sure there are other ways we could do this.  

In any event, I would prefer not to tackle making further foundational changes to the type inspection infrastructure in this PR, which is already quite big.  We could dive into this deeper in a follow up PR.

Reply:
I like the "task scheme" idea, agree it shouldn't be done here.

I didn't think about being about to load other implementations from other service scopes into the list at runtime, which may be defined in other projects that aren't direct dependencies - that's a good reason to use this.